### DISCUSSION TITLE
real number outputs

### DISCUSSION THREAD

**Discussion by WATLE** [softwareQinc/staq#42] [General]
This program produces the following output
U(pi/2,0,pi) q[2];
U(0,0,pi/4) q[2];
CX q[3],q[2];
U(0,0,-pi/4) q[2];
U(0,0,pi/4) q[1];...

I want the following output
U(1.5707963267949,0,3.14159274101257) q[2];
U(0,0,0.785398185253143) q[2];
CX q[3],q[2];
U(0,0,-0.785398185253143) q[2];
U(0,0,0.785398185253143) q[1];

I made "qelib2.inc" .
qelib2.inc is placed in the same location as the qasm file
I used qelib2.inc instead of "qelib1.inc".

The first 11 lines of qelib2 are:
// Quantum Experience (QE) Standard Header
// file: qelib2.inc

// --- QE Hardware primitives ---

// 3-parameter 2-pulse single qubit gate
gate u3(theta,phi,lambda) q { U(theta+1.0-1.0,phi+1.0-1.0,lambda+1.0-1.0) q; }
// 2-parameter 1-pulse single qubit gate
gate u2(phi,lambda) q { U(pi/2+1.0-1.0,phi+1.0-1.0,lambda+1.0-1.0) q; }
// 1-parameter 0-pulse single qubit gate
gate u1(lambda) q { U(0.0,0.0,lambda+1.0-1.0) q; }

The rest is the same as qelib1.inc

In this way I got a real output.
Because pi + 1.0-1.0 = 3.14159 ...
Note: If the original file has a pi in the U function, it won't work as it won't be converted.

**Comment by WATLE** [comment]
Note: If device have one-way CX gate, staq will automatically use Hadamard gate .
This method does not work for that gate.

**Comment by vsoftco** [comment]
Thanks @WATLE , we kept it symbolic whenever possible, so we can do more aggressive optimizations. Would it be useful to have a compiler option that evaluates everything? If yes, can you please tell us a use case?

**Reply by WATLE** [reply]
Evaluates everything option is very useful.
When reading the output with another program, it is necessary to parse it unless it is in the form of a real number.

**Comment by 525125** [comment]
We've added the staq option `--evaluate-all`. If you do it with `--map-to-device` you may need to run a `--simplify` pass after mapping.

**Reply by WATLE** [reply]
Thank you!

**Reply by WATLE** [reply]
In staq, H gate is used to switch the control and target of CNOT gate. Is --evaluate-all applied to it?

**Reply by vsoftco** [reply]
adding @525125 to the thread

**Reply by vsoftco** [reply]
@WATLE Done, see https://github.com/softwareQinc/staq/commit/e270640491c4ef3eb45cf8c88925f27c4e184f00

**Reply by WATLE** [reply]
thank you!
